### PR TITLE
Support keyword arguments for qjit-compiled functions

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**

Currently in Catalyst, keyword arguments to functions are not supported by qjit. for example: 

```
@qjit()
def f(x, y):
    return x * y

result = f(3, y=2)
assert result == f(2, 3) 
```
would result in the following exception raised
```
TypeError: fn() missing 2 required positional arguments: 'x' and 'y'
```

The main reason for this is that `kwargs` is no passed to `jit_compile` function.

**Description of the Change:**
We pass `kwargs` down to `jit_compile` and then to `capture` and finally to `trace_to_jaxpr` functions and use it to modify the tracers. 
We also added support for grad, jacobian, and value_and_grad. 
vmap already worked fine (note that vmap uses jax.vmap when outside qjit which didn't work with kwargs when I tested). jvp/vjp dib't support kwargs which is the same behaviour as the jax versions.

**Benefits:**
User can pass keyword arguments when using qjit. 

**Possible Drawbacks:**
`static_argnums` is still not supported when using keyword arguments. To support this, we need to first support `static_argnames` to be able to filter static `kwargs`

**Related GitHub Issues:**
[sc-67133]

### CODE REVIEW

**Comment by dime10** [comment]
Thanks @mehrdad2m, this sounds great! I have two questions:
- Are we ensuring kwargs are passed in the right order to compiled functions? Since kwargs can be out of order I wonder if this is consistent compared to passing those same arguments positionally (in the right order).
- I think the SC ticket includes all public API functions, have had a chance to confirm the level of support for kwargs in those? (thinking along the lines of vmap, grad, and all those functional operators)

**Comment by mehrdad2m** [comment]
- > Thanks @mehrdad2m, this sounds great! I have two questions:
> 
> * Are we ensuring kwargs are passed in the right order to compiled functions? Since kwargs can be out of order I wonder if this is consistent compared to passing those same arguments positionally (in the right order).
> * I think the SC ticket includes all public API functions, have had a chance to confirm the level of support for kwargs in those? (thinking along the lines of vmap, grad, and all those functional operators)



> Thanks @mehrdad2m, this sounds great! I have two questions:
> 
> * Are we ensuring kwargs are passed in the right order to compiled functions? Since kwargs can be out of order I wonder if this is consistent compared to passing those same arguments positionally (in the right order).
> * I think the SC ticket includes all public API functions, have had a chance to confirm the level of support for kwargs in those? (thinking along the lines of vmap, grad, and all those functional operators)

Thanks @dime10 for the review. 

- Regarding the ordering I think it preserves the order of kwargs correctly, but I will double check and add a test. 
- I didn't notice that I need to check the other functions. I will do that and add tests as well.
WIll let you know when it is finished

File: frontend/catalyst/api_extensions/differentiation.py [hunk 1]
```diff
@@ -597,21 +597,22 @@ def __call__(self, *args, **kwargs):
...
                     self.grad_params.scalar_out,
                     self.grad_params.h,
                     self.grad_params.argnum,
                     len(args_data),
-                    in_tree,
+                    in_arg_tree,
                     self.grad_params.with_value,
                 )
-                jaxpr, out_tree = _make_jaxpr_check_differentiable(fn, grad_params, *args)
+                input_data_flat, _ = tree_flatten((args, kwargs))
```
Comment (rmoyard) [inline frontend/catalyst/api_extensions/differentiation.py:610]: Is there not a way to merge this tree flatten with the one above?

File: frontend/catalyst/jit.py [hunk 2]
```diff
@@ -678,7 +680,7 @@ def run(self, args, kwargs):
             Any: results of the execution arranged into the original function's output PyTrees
         """
 
-        results = self.compiled_function(*args, **kwargs)
+        results = self.compiled_function(*(*args, kwargs), {})
```
Comment (rmoyard) [inline frontend/catalyst/jit.py:683]: Not sure I fully understand this one? Should not *args, **kwargs work? If not, don't we need ** in front of the kwargs?

File: frontend/catalyst/qfunc.py [hunk 3]
```diff
@@ -151,17 +151,17 @@ def _eval_quantum(*args):
 
             out_tree_expected.append(out_tree_exp)
             dynamic_args = filter_static_args(args, static_argnums)
```
Comment (rmoyard) [inline frontend/catalyst/qfunc.py:153]: This filtering is not needed for kwargs?

File: frontend/test/pytest/test_gradient.py [hunk 4]
```diff
@@ -1200,6 +1200,64 @@ def circuit(phi, psi):
...
 
 
+def test_calssical_kwargs():
+    """Test the gradient on a classical function with keyword arguments"""
+
+    @qjit
+    def f1(x, y, z):
+        return x * (y - z)
+
+    result = qjit(grad(f1, argnum=0))(3.0, y=1.0, z=2.0)
```
Comment (rmoyard) [inline frontend/test/pytest/test_gradient.py:1210]: Do you what happens if argnum refers to a non positional argument? Not sure what happens in Jax neither

**Review by rmoyard — COMMENTED** [review]
Great job :+1: I have left some questions. I think we might need to change `vjp` and `jvp` to work with kwargs but it is probably out of scope?

**Comment by mehrdad2m** [comment]
> Great job 👍 I have left some questions. I think we might need to change `vjp` and `jvp` to work with kwargs but it is probably out of scope?

Thanks @rmoyard for the review. I checked the jvp/vjp and I think jax doesn't support kwargs, so I don't know if we want to add this support for catalyat or not. Also in the current format, we pass the function arguments in a list as the second argument of jvp (`params`). so it is a bit different that the traditional (`*args, **kwargs)`

**Comment by rmoyard** [comment]
No worries, @mehrdad2m thanks for checking!

File: frontend/catalyst/jit.py (hunk 2 above)
Comment (mehrdad2m) [reply frontend/catalyst/jit.py]: Actually, after a second look. I could make it work without this change by changing the call funciton of CompiledFunction class.

File: frontend/catalyst/api_extensions/differentiation.py (hunk 1 above)
Comment (rmoyard) [reply frontend/catalyst/api_extensions/differentiation.py:610]: Thanks

**Review by rmoyard — APPROVED** [review]
Thanks @mehrdad2m it looks good to me :+1:

**Comment by mehrdad2m** [comment]
> Thanks @mehrdad2m, this sounds great! I have two questions:
> 
> * Are we ensuring kwargs are passed in the right order to compiled functions? Since kwargs can be out of order I wonder if this is consistent compared to passing those same arguments positionally (in the right order).
> * I think the SC ticket includes all public API functions, have had a chance to confirm the level of support for kwargs in those? (thinking along the lines of vmap, grad, and all those functional operators)

Hi @dime10, Just a summary of what I have checked so far. I added support for grad, jacobian, and value_and_grad. vmap worked fine (not that vmap uses jax.vmap when outside qjit which didn't work with kwargs when I tested). jvp/vjp dob't support kwargs which is the same behaviour as the jax versions.
Is there anything that I have missed?

**Comment by dime10** [comment]
> > Thanks @mehrdad2m, this sounds great! I have two questions:
> > 
> > * Are we ensuring kwargs are passed in the right order to compiled functions? Since kwargs can be out of order I wonder if this is consistent compared to passing those same arguments positionally (in the right order).
> > * I think the SC ticket includes all public API functions, have had a chance to confirm the level of support for kwargs in those? (thinking along the lines of vmap, grad, and all those functional operators)
> 
> Hi @dime10, Just a summary of what I have checked so far. I added support for grad, jacobian, and value_and_grad. vmap worked fine (not that vmap uses jax.vmap when outside qjit which didn't work with kwargs when I tested). jvp/vjp dob't support kwargs which is the same behaviour as the jax versions. Is there anything that I have missed?

Sounds great, thank you!

**Pull request merged**